tests/integration/dump/test_dump_dbms.py

```diff
@@ -111,65 +116,247 @@ def path_size(dump_path):
+                times_indexed += 1
+
+                # assert data dumped is correct
+                for pea_id in range(shards):
+                    assert_dump_data(dump_path, full_docs, shards, pea_id)
+
+
+def _test_dump_prepare(emb_size, nr_docs, run_basic, shards, tmpdir):
     docs = list(get_documents(nr=nr_docs, index_start=0, emb_size=emb_size))
     assert len(docs) == nr_docs
-    nr_search = 1
+    nr_search = 3
 
     os.environ['USES_AFTER'] = '_merge_matches' if shards > 1 else '_pass'
-    os.environ['SHARDS'] = str(shards)
-
-    def _validate_results_nonempty(resp):
-        assert len(resp.docs) == nr_search
-        for d in resp.docs:
-            if nr_docs < 10:
-                assert len(d.matches) == nr_docs
-            else:
-                # TODO does it return all of them no matter how many?
-                assert len(d.matches) > 0
-            for m in d.matches:
-                assert m.embedding.shape[0] == emb_size
-                assert (
-                    DBMSIndexDriver._doc_without_embedding(m).SerializeToString()
-                    is not None
-                )
-                assert 'hello world' in m.text
-                assert f'tag data' in m.tags['tag_field']
+    os.environ['QUERY_SHARDS'] = str(shards)
 
-    def error_callback(resp):
-        raise Exception('error callback called')
+    validation_query = functools.partial(
+        _validate_results_nonempty, nr_search, nr_docs * 2, emb_size
+    )  # x 2 because we run it twice
 
     if run_basic:
-        basic_benchmark(
-            tmpdir, docs, _validate_results_nonempty, error_callback, nr_search
-        )
+        basic_benchmark(tmpdir, docs, validation_query, _error_callback, nr_search)
+
+    dump_path = os.path.join(str(tmpdir), 'dump_dir')
+    os.environ['DBMS_WORKSPACE'] = os.path.join(str(tmpdir), 'index_ws')
+    os.environ['QUERY_WORKSPACE'] = os.path.join(str(tmpdir), 'query_ws')
+
+    return validation_query, docs, dump_path, nr_search
+
+
+def _assert_order_ops(ops_log, ops: List[str]):
+    print(ops_log)
+    assert len(ops_log) > 0
+    last_idx_found = -1
+    for rec in ops_log:
+        for i, op in enumerate(ops):
+            if op in rec:
+                print(f'found {op} in {rec}')
+                if i != last_idx_found + 1:
+                    return False
+                last_idx_found = i
+    return last_idx_found == len(ops) - 1
+
+
+# log of the statements in the threading example
+# to assert order
+operations = []
+
+
+def _print_and_append_to_ops(statement):
+    global operations
+    operations.append(statement)
+    print(statement, flush=True)
+
+
+@pytest.mark.repeat(5)
+@pytest.mark.parametrize('nr_docs', [700])
+@pytest.mark.parametrize('emb_size', [10])
+def test_threading_query_while_reloading(tmpdir, nr_docs, emb_size, mocker):
+    global operations
+
+    # TODO better way to test async procedure call order
+    # patch
+    def _rolling_update(self, dump_path):
```

Comment:
is this needed?

Reply:
You mean the test in general, or the patched method? 

If the former, maybe not anymore, since we've added the stress test.

If the latter, I patch it in order to log and then make sure the operations are run in a specific order

Reply:
the patch, ssems weird, but okey.

Reply:
Maybe it is better to remove it, as it is quite fragile. I wrote it before we added the more comprehensive jinad threading stress test, to show that is possible to query while reloading.